tests/Basics/SupervisedBufferTest.cpp

```diff
@@ -0,0 +1,479 @@
+  auto& global = GlobalResourceMonitor::instance();
+  ResourceMonitor monitor{global};
+
+  {
+    ResourceUsageScope usageScope(monitor);
+    AqlValue largeValue;
+    {
+      Builder builder;
+      builder.openArray();
+      builder.add(Value(std::string(1024, 'a')));
+      builder.close();
+
+      ASSERT_EQ(monitor.current(), 0);
+      largeValue = AqlValue{builder.slice(), builder.size()};
+      monitor.increaseMemoryUsage(largeValue.memoryUsage());
+
+      ASSERT_EQ(monitor.current(), largeValue.memoryUsage());
+    }
+    ASSERT_EQ(monitor.current(), largeValue.memoryUsage());
+
+    monitor.decreaseMemoryUsage(largeValue.memoryUsage());
+    largeValue.destroy();
+    ASSERT_EQ(monitor.current(), 0);
+  }
+
+  {
+    ResourceUsageScope usageScope(monitor);
+    AqlValue smallValue;
+    {
+      Builder builder;
+      builder.openArray();
+      builder.add(Value(1));
+      builder.close();
+      ASSERT_EQ(monitor.current(), 0);
+      smallValue = AqlValue{builder.slice(), builder.size()};
+      // memoryUsage for small values  is zero
+      // because they fit in the AqlValue's inline buffer
+      monitor.increaseMemoryUsage(smallValue.memoryUsage());
+      ASSERT_EQ(monitor.current(), smallValue.memoryUsage());
+    }
+    // is the same as comparing to smallValue.memoryUsage
+    ASSERT_EQ(monitor.current(), 0);
+
+    monitor.decreaseMemoryUsage(smallValue.memoryUsage());
+    smallValue.destroy();
+    ASSERT_EQ(monitor.current(), 0);
+  }
+}
+
+TEST(SupervisedbufferTest, AccountsMemoryLargeAndSmallValuesSupervisedBuffer) {
+  auto& global = GlobalResourceMonitor::instance();
+  ResourceMonitor monitor{global};
+
+  {
+    ResourceUsageScope usageScope(monitor);
+    AqlValue largeValue;
+    {
+      SupervisedBuffer supervisedBuffer(monitor);
+      Builder builder(supervisedBuffer);
+      builder.openArray();
+      builder.add(Value(std::string(1024, 'a')));
+      builder.close();
+
+      ASSERT_GT(monitor.current(), 0);
+      largeValue = AqlValue{builder.slice(), builder.size()};
+      // the buffer has already increased the monitor for its
+      // capacity, this call adds the AqlValue footprint on top.
+      monitor.increaseMemoryUsage(largeValue.memoryUsage());
+      // While builder exists, monitor >= aql value usage (buffer + aql value)
+      ASSERT_GE(monitor.current(), largeValue.memoryUsage());
+    }
+    // now monitor is only the aql value usage
+    ASSERT_EQ(monitor.current(), largeValue.memoryUsage());
+
+    monitor.decreaseMemoryUsage(largeValue.memoryUsage());
+    largeValue.destroy();
+    ASSERT_EQ(monitor.current(), 0);
+  }
+
+  {
+    ResourceUsageScope usageScope(monitor);
+    AqlValue smallValue;
+    {
+      SupervisedBuffer supervisedBuffer(monitor);
+      Builder builder(supervisedBuffer);
+      builder.openArray();
+      builder.add(Value(1));
+      builder.close();
+      ASSERT_GT(monitor.current(), 0);
```

Comment:
I think we can actually assert the following here:
1) Before writing into the buffer take "current" value, add this small entry in, then assert monitor is still on current (first into _local place)
2) When reading the first current assert > 0